File: src-tauri/src/commands/files.rs
use crate::db::files::{self, FileEntity};
use crate::db::DbState;
use crate::error::{AppError, AppResult};
use std::path::PathBuf;
use tauri::{AppHandle, Manager, State};

#[tauri::command]
pub fn import_file(
    app: AppHandle,
    state: State<DbState>,
    space_id: String,
    source_path: String,
) -> AppResult<FileEntity> {
    let files_dir: PathBuf = app
        .path()
        .app_data_dir()
        .map_err(|e| AppError::Db(e.to_string()))?
        .join("files");
    let conn = state.0.lock().unwrap();
    files::import_file(
        &conn,
        &files_dir,
        space_id,
        std::path::Path::new(&source_path),
    )
}

#[tauri::command]
pub fn create_file_link(
    state: State<DbState>,
    space_id: String,
    title: String,
    url: String,
) -> AppResult<FileEntity> {
    let conn = state.0.lock().unwrap();
    files::create_file_link(&conn, space_id, title, url)
}

#[tauri::command]
pub fn list_files(state: State<DbState>, space_id: String) -> AppResult<Vec<FileEntity>> {
    let conn = state.0.lock().unwrap();
    files::list_files(&conn, &space_id)
}

#[tauri::command]
pub fn get_file(state: State<DbState>, entity_id: String) -> AppResult<FileEntity> {
    let conn = state.0.lock().unwrap();
    files::get_file(&conn, &entity_id)
}

/// Copies an imported file out of Nookly's storage to `destination`, e.g. a path
/// picked in a native save dialog. Link-only files have nothing local to copy.
#[tauri::command]
pub fn export_file(state: State<DbState>, entity_id: String, destination: String) -> AppResult<()> {
    let file = {
        let conn = state.0.lock().unwrap();
        files::get_file(&conn, &entity_id)?
    };
    let source = file.local_path.ok_or_else(|| {
        AppError::InvalidInput("this file is a link, there is no local copy to export".into())
    })?;
    std::fs::copy(&source, &destination)
        .map(|_| ())
        .map_err(|err| AppError::Io(err.to_string()))
}

File: packages/frontend/src/lib/api/files.ts
import { invoke } from "@tauri-apps/api/core";
import type { FileEntity } from "./types";

export function importFile(spaceId: string, sourcePath: string): Promise<FileEntity> {
  return invoke("import_file", { spaceId, sourcePath });
}

/// What a link turned into: a stored File, or a webpage to save as a Bookmark.
export type LinkImport = { kind: "file"; file: FileEntity } | { kind: "webpage" };

/// Downloads the file behind a link (share links included) into storage.
export function importFileFromUrl(spaceId: string, url: string): Promise<LinkImport> {
  return invoke("import_file_from_url", { spaceId, url });
}

/// Downloads the file behind a link-only File, from before links were stored.
export function downloadLinkedFile(entityId: string): Promise<LinkImport> {
  return invoke("download_linked_file", { entityId });
}

export function getFile(entityId: string): Promise<FileEntity> {
  return invoke("get_file", { entityId });
}

export function listFiles(spaceId: string): Promise<FileEntity[]> {
  return invoke("list_files", { spaceId });
}

/// Copies an imported file out of Nookly's storage to `destination`.
export function exportFile(entityId: string, destination: string): Promise<void> {
  return invoke("export_file", { entityId, destination });
}

/// Swaps the stored copy for a newer version; same File, labels and relationships.
export function replaceFile(entityId: string, sourcePath: string): Promise<FileEntity> {
  return invoke("replace_file", { entityId, sourcePath });
}

/// Corrects a File's Added date; it otherwise defaults to the day it was
/// imported or downloaded.
export function setFileAddedAt(entityId: string, addedAt: string): Promise<FileEntity> {
  return invoke("set_file_added_at", { entityId, addedAt });
}

/// Adds a file from disk by reference, leaving it where it is.
export function referenceFile(spaceId: string, path: string): Promise<FileEntity> {
  return invoke("reference_file", { spaceId, path });
}

/// Copies a referenced file into Nookly's own storage.
export function copyFileIntoStorage(entityId: string): Promise<FileEntity> {
  return invoke("copy_file_into_storage", { entityId });
}

/// Opens the file (stored copy or referenced original) in its default app.
export function openFile(entityId: string): Promise<void> {
  return invoke("open_file", { entityId });
}

/// Shows the file in Finder or the platform's file manager.
export function revealFile(entityId: string): Promise<void> {
  return invoke("reveal_file", { entityId });
}

export interface OpenWithApp {
  name: string;
  path: string;
  isDefault: boolean;
  /// PNG data URL, when the platform provides one.
  icon: string | null;
}

/// Apps the system says can open this file, its default app first. Empty where
/// the platform can't be asked (only macOS so far).
export function listOpenWithApps(entityId: string): Promise<OpenWithApp[]> {
  return invoke("list_open_with_apps", { entityId });
}

/// Opens the file with a specific app, by name or path.
export function openFileWith(entityId: string, appPath: string): Promise<void> {
  return invoke("open_file_with", { entityId, appPath });
}
